pulsar-broker-auth-sasl/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderSasl.java

```diff
@@ -109,4 +128,160 @@ public AuthenticationState newAuthState(AuthData authData,
             throw new AuthenticationException(t.getMessage());
         }
     }
+
+    // for http auth.
+    private static final long LIVE_SECONDS = 3600;
+    // A signer for http role token, with random secret.
+    private SaslRoleTokenSigner signer;
+
+    /**
+     * Returns null if authentication has not completed.
+     * Return auth role if authentication has completed, and httpRequest's role token contains the authRole
+     */
+    public String authRoleFromHttpRequest(HttpServletRequest httpRequest) throws AuthenticationException {
+        String tokenStr = httpRequest.getHeader(SASL_AUTH_ROLE_TOKEN);
+
+        if (tokenStr == null) {
+            return null;
+        }
+
+        String unSigned = signer.verifyAndExtract(tokenStr);
+        SaslRoleToken token;
+
+        try {
+            token = SaslRoleToken.parse(unSigned);
+            if (log.isDebugEnabled()) {
+                log.debug("server side get role token: {}, session in token:{}, session in request:{}",
+                    token, token.getSession(), httpRequest.getRemoteAddr());
+            }
+        } catch (Exception e) {
+            log.error("token parse failed, with exception: ",  e);
+            return SASL_AUTH_ROLE_TOKEN_EXPIRED;
+        }
+
+        if (!token.isExpired()) {
+            return token.getUserRole();
+        } else if (token.isExpired()) {
+            return SASL_AUTH_ROLE_TOKEN_EXPIRED;
+        } else {
+            return null;
+        }
+    }
+
+    private String createAuthRoleToken(String role, String sessionId) {
+        long expires = System.currentTimeMillis() + LIVE_SECONDS * 1000; // 1 hour
+        SaslRoleToken token = new SaslRoleToken(role, sessionId, expires);
+
+        String signed = signer.sign(token.toString());
+        if (log.isDebugEnabled()) {
+            log.debug("create role token token: {}, role: {} session :{}, expires:{}\nsigned:{}",
+                token, token.getUserRole(), token.getSession(), token.getExpires(), signed);
+        }
+        return signed;
+    }
+
+    private ConcurrentHashMap<Long, AuthenticationState> authStates = new ConcurrentHashMap<>();
+
+    // return authState if it is in cache.
+    private AuthenticationState getAuthState(HttpServletRequest request) {
+        String id = request.getHeader(SASL_STATE_SERVER);
+        if (id == null) {
+            return null;
+        }
+        return authStates.get(new Long(id));
```

Comment:
use `Long.parse` to parse a long number from a string. You might need to catch NumberFormatException as well.